libs/ImpSpec/src/Test/ImpSpec/Internal.hs

```diff
@@ -0,0 +1,388 @@
+import Prettyprinter (
+  Doc,
+  Pretty (..),
+  annotate,
+  defaultLayoutOptions,
+  hcat,
+  indent,
+  layoutPretty,
+  line,
+  vsep,
+ )
+import Prettyprinter.Render.Terminal (AnsiStyle, Color (..), color, renderLazy)
+import System.Random (randomR, split)
+import System.Random.Stateful (IOGenM, applyIOGen, newIOGenM)
+import Test.HUnit.Lang (FailureReason (..), HUnitFailure (..))
+import Test.Hspec (Spec, SpecWith, beforeAll, beforeAllWith)
+import Test.Hspec.Core.Spec (
+  Example (..),
+  Result (..),
+  paramsQuickCheckArgs,
+ )
+import qualified Test.Hspec.Core.Spec as H
+import Test.ImpSpec.Expectations
+import Test.ImpSpec.Random
+import Test.QuickCheck (Arbitrary, Args (chatty, replay), Testable (..), counterexample, ioProperty)
+import Test.QuickCheck.Gen (Gen (..))
+import Test.QuickCheck.GenT (MonadGen (..))
+import Test.QuickCheck.Random (QCGen (..), integerVariant, mkQCGen)
+import UnliftIO (MonadIO (liftIO), MonadUnliftIO (..))
+import UnliftIO.Exception (
+  Exception (..),
+  SomeException (..),
+  catchAny,
+  catchAnyDeep,
+  throwIO,
+ )
+import UnliftIO.IORef
+
+data ImpState t = ImpState
+  { impStateSpecState :: !(ImpSpecState t)
+  , impStateLog :: !(Doc AnsiStyle)
+  }
+
+data ImpEnv t = ImpEnv
+  { impEnvSpecEnv :: !(ImpSpecEnv t)
+  , impEnvStateRef :: !(IORef (ImpState t))
+  , impEnvQCGenRef :: !(IOGenM QCGen)
+  , impEnvQCSize :: !Int
+  }
+
+class ImpSpec t where
+  type ImpSpecEnv t = (r :: Type) | r -> t
+  type ImpSpecEnv t = Proxy t
+  type ImpSpecState t = (r :: Type) | r -> t
+  type ImpSpecState t = Proxy t
+
+  impInitIO :: QCGen -> IO (ImpInit t)
+  default impInitIO :: (ImpSpecEnv t ~ Proxy t, ImpSpecState t ~ Proxy t) => QCGen -> IO (ImpInit t)
+  impInitIO _ = pure $ ImpInit Proxy Proxy
+
+  -- | This will be the very first action that will run in all `ImpM` specs.
+  impPrepAction :: ImpM t ()
+  impPrepAction = pure ()
+
+data ImpInit t = ImpInit
+  { impInitEnv :: ImpSpecEnv t
+  , impInitState :: ImpSpecState t
+  }
+deriving instance (Eq (ImpSpecEnv t), Eq (ImpSpecState t)) => Eq (ImpInit t)
+deriving instance (Ord (ImpSpecEnv t), Ord (ImpSpecState t)) => Ord (ImpInit t)
+deriving instance (Show (ImpSpecEnv t), Show (ImpSpecState t)) => Show (ImpInit t)
+
+-- | Stores extra information about the failure of the unit test
+data ImpException = ImpException
+  { ieAnnotation :: [Doc AnsiStyle]
+  -- ^ Description of the IO action that caused the failure
+  , ieThrownException :: SomeException
+  -- ^ Exception that caused the test to fail
+  }
+  deriving (Show)
+
+instance Exception ImpException where
+  displayException = ansiDocToString . prettyImpException
+
+prettyImpException :: ImpException -> Doc AnsiStyle
+prettyImpException (ImpException ann e) =
+  vsep $
+    mconcat
+      [ ["Annotations:"]
+      , zipWith indent [0, 2 ..] ann
+      , ["Failed with Exception:", indent 4 $ pretty (displayException e)]
+      ]
+
+newtype ImpM t a = ImpM {unImpM :: ReaderT (ImpEnv t) IO a}
```

Comment:
It seems at first that it work, but:
* it would lead to terrible type signatures on all functions (i.e. `impTest :: ImpM (ImpLedgerEnv era) (ImpLedgerState era) ()`).
* it would not be as future extensible as current approach (eg. adding potentially other things akin to environment or state, like events of some sorts for instance)
* it would make the whole initialization much uglier:
```haskell
class ImpSpec e s where

  impInitIO :: QCGen -> IO (ImpInit e s t)
```
* Most importantly we would not have a way to implement this kind of functionality:
```haskell
class ImpSpec t where
  impPrepAction :: ImpM t ()
```
  In other words there would be no way to customize Imp functionality that does not depend on neither state nor environment.

Those are just facts that immediately come to my mind, but in general for any Haskell interface, if it is possible to reduce number of type parameters then it is a good idea to do so.